### PR TITLE
dw stack blog

### PR DESCRIPTION
**Pull request by ChuckHend**

### CODE REVIEW

**Review by DarrenBaldwin07 — APPROVED** [review]
Looks awesome!!

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+Developing a foreign data wrapper with Wrappers is straight-forward. You simply implement the [ForeignDataWrapper](https://github.com/supabase/wrappers/blob/main/supabase-wrappers/src/interface.rs) [trait](https://doc.rust-lang.org/book/ch10-02-traits.html) in your Rust extension. Implementing a trait in Rust is a lot like implementing an abstract base class in Python or an interface in Java, and ultimately, the the functions in the trait are what map your SQL query into the requests that get sent to the foreign data server. In the case of clerk_fdw, we need to map a query like `select * from users` into a request to the Clerk API, and that is exactly what happens in the trait functions.
```
Comment (ChuckHend) [inline docusaurus/blog/2024-01-15-data-warehouse-stack/index.md:38]: @Jayko001 , can you provide any feedback on this section?

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+In fact, we built our internal data warehouse at Tembo using this Stack. We have data spread across several different vendors and systems, but want a single place to query, analyze, and visualize all of our data. Tembo has operational data in a few different locations, and we used foreign data wrappers to connect Postgres to multiple external sources, partitioning to improve performance and set retention policies, and a built-in scheduler to keep our data up to date.
+
+## Loading data from several sources
+
+![task](./task.png 'task')
+
+To build our data warehouse at Tembo, we needed to pull our operational data into Postgres from several external sources, including other Postgres instances:
+
+- Postgres - we run a dedicated Postgres cluster for the Tembo Cloud backend. We refer to this as our "control-plane". This is where we maintain all the metadata for customer;'s Tembo instances. For example, what cpu, memory, storage, when their instance was created, what its name and the organization it belongs to, etc. We also run a message queue in Postgres, which contains an archive of historical events from the system.
```
Comment (Jayko001) [inline docusaurus/blog/2024-01-15-data-warehouse-stack/index.md:18]: typo at `metadata for customer;'s`

Comment (ChuckHend) [reply docusaurus/blog/2024-01-15-data-warehouse-stack/index.md]: woops!

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md (hunk 1 above)
Comment (Jayko001) [reply docusaurus/blog/2024-01-15-data-warehouse-stack/index.md:38]: Yeah, it look good.

btw, typo at `Java, and ultimately, the the functions`

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+$function$
+```
+
+Then, we set up the pg_cron job to call this function as frequently as we need:
+
+```sql
+SELECT cron.schedule('update-clusters', '5 seconds', 'CALL refresh_clusters()');
+```
+
+We can easily check and see what jobs we've scheduled in our data warehouse by peeking at he `cron.job` table, and its easy to interpret.
```
Comment (Jayko001) [inline docusaurus/blog/2024-01-15-data-warehouse-stack/index.md:155]: typo at `by peeking at he `cron.job` table`

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+The majority of our dashboard queries aggregate data over time, and most commonly on a daily interval. So, we can partition our tables by day, and only query the partitions we need to answer our questions. This provides a substantial improvement to the performance of those queries which makes our dashboards very snappy.
+
+Our stakeholders do not require visualization for the entirety of our metric data, in fact they are typically only concerned with a 30 days at most. Therefore, we only need to retail 30 days in our data warehouse storage. By setting up a retention policy, we can automatically drop partitions that are older than 30 days, and reclaim that storage. Dropping a partition is much faster than deleting rows from a table. As we'll see in a moment, it is trivial to configure partitioning on Postgres if you use [pg_partman](https://github.com/pgpartman/pg_partman) (which is my personal favorite Postgres extensions). Without pg_partman, it is up to you to handle the creation and deletion of partitions.
```
Comment (Jayko001) [inline docusaurus/blog/2024-01-15-data-warehouse-stack/index.md:187]: typo at `to retail 30 days in our`

Comment (ChuckHend) [reply docusaurus/blog/2024-01-15-data-warehouse-stack/index.md]: https://github.com/tembo-io/website/pull/279/commits/626f05f288a1b5c923caa2e8310571c925acfe52

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md (hunk 3 above)
Comment (ChuckHend) [reply docusaurus/blog/2024-01-15-data-warehouse-stack/index.md]: https://github.com/tembo-io/website/pull/279/commits/626f05f288a1b5c923caa2e8310571c925acfe52

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md (hunk 1 above)
Comment (ChuckHend) [reply docusaurus/blog/2024-01-15-data-warehouse-stack/index.md]: https://github.com/tembo-io/website/pull/279/commits/626f05f288a1b5c923caa2e8310571c925acfe52

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
+---
+slug: data-warehouse-stack
+title: 'Data Warehouse Stack on Tembo Cloud'
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-15-data-warehouse-stack/index.md:3]: I know you will change this, but just noting :)

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+authors: [adam]
+tags: [postgres, extensions, stacks, data-warehouse]
+---
+
+Building a data warehouse is a complex task. Typically, a data engineering team would need to integrate several different tools and technologies to bring data into a central location from various data sources. Tembo reduces the complexity associated with building a data warehouse by providing a pre-configured Stack that includes all the tools you need to build a data warehouse without running additional infrastructure, all accessible to you from within Postgres.
+
+In fact, we built our internal data warehouse at Tembo using this Stack. We have data spread across several different vendors and systems, but want a single place to query, analyze, and visualize all of our data. Tembo has operational data in a few different locations, and we used foreign data wrappers to connect Postgres to multiple external sources, partitioning to improve performance and set retention policies, and a built-in scheduler to keep our data up to date.
+
+## Loading data from several sources
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-15-data-warehouse-stack/index.md:12]: I'd start with the "Task". What was the expected end goal, why we did all this?

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+- [Clerk.dev](https://clerk.com/) - we partner with Clerk to provide authentication and authorization for all our customers, as well as management of our user's organizations.
+
+Our task was to bring all this data into a single place so that we could join the data together, analyze it, and present it in our dashboards.
+
+Moving data from several different sources into a single place is a common task for data engineers. There are lots of tools in the ecosystem to help with this task. Many organizations bring in external tools, and vendors to handle this orchestration. Tools like [Airflow](https://airflow.apache.org/), [dbt](https://github.com/dbt-labs/dbt-core), [Fivetran](https://www.fivetran.com/blog/modern-data-warehouse), [Dagster](https://dagster.io/) are very popular, and outstanding projects, but using them comes at a huge cost. Every time we bring in a new technology into the ecosystem it expands the cognitive load on the engineering and analytics teams. Many software systems today consist of far too many tools and technologies for any human to keep in their head. Some engineers at Uber spoke briefly about this [recently](https://youtu.be/zQ5e3B5I-U0?t=81).
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-15-data-warehouse-stack/index.md:24]: Add a sub-section like "Why do it all in Postgres"

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+UPDATE part_config
+    SET retention = '30 days',
+        retention_keep_table = false,
+        retention_keep_index = false,
+        infinite_time_partitions = true
+    WHERE parent_table = 'public.metric_values';
+```
+
+We use partitioning in other places at Tembo as well, and we wrote a bit about those use-cases earlier this year in [another blog](https://tembo.io/blog/table-version-history).
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-15-data-warehouse-stack/index.md:221]: Can we add a "How it all comes together" section to wrap things up. We introduce a lot of different pieces but don't tie it up.

We should also have an end result picture from the preset dashboard but with data from September or something to show what the end result looks like.

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+## DB Objects as code with SQL Migrations
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-15-data-warehouse-stack/index.md:223]: I'd remove this for now? Seems very lightweight.

**Review by samay-sharma — COMMENTED** [review]
A few meta points on this one:

* I felt we draw too many analogies in this one. fdw is like Airflow, dagster, fivetran. trait is like .. Hive etc. etc. I don't that's trimmable.
* We spend quite a lot of time explaining each extension. I think those descriptions can be shorter. We can give a sharp quick explanation of the tool we are utilizing and then just get back to our solution building.
* There's also a lot of (some subtle, some not-subtle) tembo plugs. I think we can reduce them.

Overall, I feel this can be made shorter and a more of a solution design etc. type blog post.

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md (hunk 5 above)
Comment (ChuckHend) [reply docusaurus/blog/2024-01-15-data-warehouse-stack/index.md:3]: Do you have any suggestions?

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md (hunk 6 above)
Comment (ChuckHend) [reply docusaurus/blog/2024-01-15-data-warehouse-stack/index.md:12]: https://github.com/tembo-io/website/pull/279/commits/7bf605e2b83e6bc8974acae9463a7354a3d03c7c

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md (hunk 7 above)
Comment (ChuckHend) [reply docusaurus/blog/2024-01-15-data-warehouse-stack/index.md:24]: https://github.com/tembo-io/website/pull/279/commits/835be51025e4f95b341f936a561a6659a99ada55

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md (hunk 9 above)
Comment (ChuckHend) [reply docusaurus/blog/2024-01-15-data-warehouse-stack/index.md]: https://github.com/tembo-io/website/pull/279/commits/724e6edd18076a15062b540d0c4982458e62ad3f

File: docusaurus/blog/2024-01-15-data-warehouse-stack/index.md (hunk 8 above)
Comment (ChuckHend) [reply docusaurus/blog/2024-01-15-data-warehouse-stack/index.md:223]: The "Wrapping up" section is meant to do this. I'll take another look at that section.

**Comment by samay-sharma** [comment]
@ChuckHend : Take a look. Can you also include the obfuscated data warehouse screenshot too as we had discussed?

**Pull request closed without merging**